### PR TITLE
expression, executor, plan: rewrite builtin function cot

### PR DESCRIPTION
**Pull request by stdupp**

### CODE REVIEW

File: expression/builtin_math_test.go [hunk 1]
```diff
@@ -937,25 +937,38 @@ func (s *testEvaluatorSuite) TestTan(c *C) {
...
+		expect float64
+		isNil  bool
+		getErr bool
 	}{
-		{nil, nil},
-		{math.Pi / 4, math.Cos(math.Pi/4) / math.Sin(math.Pi/4)}, // cot pi/4 does not return 1 actually
-		{-math.Pi / 4, math.Cos(-math.Pi/4) / math.Sin(-math.Pi/4)},
-		{math.Pi * 3 / 4, math.Cos(math.Pi*3/4) / math.Sin(math.Pi*3/4)},
-		{"3.1415926", math.Cos(3.1415926) / math.Sin(3.1415926)},
+		{nil, 0, true, false},
```
Comment (shenli) [inline expression/builtin_math_test.go]: Could you add cases like math.Pi/2, math.Pi ?

Comment (stdupp) [inline expression/builtin_math_test.go]: Ok..

File: plan/typeinfer_test.go [hunk 2]
```diff
@@ -112,6 +112,7 @@ func (s *testPlanSuite) TestInferType(c *C) {
 		{"atan(c_double,c_double)", mysql.TypeDouble, charset.CharsetBin, mysql.BinaryFlag, mysql.MaxRealWidth, types.UnspecifiedLength},
 		{"asin(c_double)", mysql.TypeDouble, charset.CharsetBin, mysql.BinaryFlag, mysql.MaxRealWidth, types.UnspecifiedLength},
 		{"acos(c_double)", mysql.TypeDouble, charset.CharsetBin, mysql.BinaryFlag, mysql.MaxRealWidth, types.UnspecifiedLength},
+		{"cot(c_double)", mysql.TypeDouble, charset.CharsetBin, mysql.BinaryFlag, mysql.MaxRealWidth, types.UnspecifiedLength},
```
Comment (zz-jason) [inline plan/typeinfer_test.go]: It's better to add more test cases for other input parameter types.

File: expression/builtin_math.go [hunk 3]
```diff
@@ -972,38 +972,34 @@ type cotFunctionClass struct {
...
+	if tan != 0 {
+		cot := 1 / tan
+		if !math.IsInf(cot, 0) && !math.IsNaN(cot) {
+			return cot, false, nil
+		}
 	}
-	// Set the result to be of type float64
-	d.SetFloat64(cos / sin)
-	return d, nil
+	return 0, false, types.ErrOverflow.GenByArgs("DOUBLE", fmt.Sprintf("cot(%s)", strconv.FormatFloat(val, 'f', -1, 64)))
```
Comment (zz-jason) [inline expression/builtin_math.go]: Add test case for this: `cot(0)` in `executor/executor_test.go`

File: expression/builtin_math_test.go [hunk 4]
```diff
@@ -937,25 +937,38 @@ func (s *testEvaluatorSuite) TestTan(c *C) {
...
-		v, err := f.eval(nil)
+	for _, test := range tests {
+		f, err := newFunctionForTest(s.ctx, ast.Cot, primitiveValsToConstants([]interface{}{test.args})...)
 		c.Assert(err, IsNil)
-		c.Log(t)
-		c.Assert(v, testutil.DatumEquals, t["Ret"][0])
+
+		result, err := f.Eval(nil)
+		if test.getErr {
+			c.Assert(err, NotNil)
```
Comment (zz-jason) [inline expression/builtin_math_test.go]: Could you add another test to check the error message ?

**Comment by stdupp** [comment]
PTAL @zz-jason

**Comment by zz-jason** [comment]
@shenli @XuHuaiyu PTAL

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@stdupp Thanks!

**Pull request merged**